### PR TITLE
Add gradle wrapper to new project template.

### LINKED ISSUE #8131: Flutter tool should be able to tell me if I don't have gradle installed
Following the guide for setting Flutter up on a linux laptop, I get the following error without gradle installed.

It would be nice to get a pointer on how to set it up, similar to what we do with the Android SDK.  From what I can see, the extra step is reasonably simple -- add `{AndroidStudio install dir}/gradle/gradle-2.14.1/bin/` to the PATH.

Error below:
➜  flutter urn
Launching lib/main.dart on Nexus 5X in debug mode...
Exception from flutter run: Exception: Unable to locate gradle. Please configure the path to gradle using 'flutter config --gradle-dir'.
package:flutter_tools/src/base/common.dart 29              throwToolExit
package:flutter_tools/src/android/gradle.dart 141          ensureGradlew.<async>
===== asynchronous gap ===========================
dart:async                                                 _Completer.completeError
package:flutter_tools/src/android/gradle.dart 184          ensureGradlew.<async>
===== asynchronous gap ===========================
dart:async                                                 Future.Future.microtask
package:flutter_tools/src/android/gradle.dart              ensureGradlew
package:flutter_tools/src/android/gradle.dart 202          buildGradleProject.<async>
===== asynchronous gap ===========================
dart:async                                                 Future.Future.microtask
package:flutter_tools/src/android/gradle.dart              buildGradleProject
package:flutter_tools/src/commands/build_apk.dart 610      buildAndroidWithGradle.<async>
===== asynchronous gap ===========================
dart:async                                                 Future.Future.microtask
package:flutter_tools/src/commands/build_apk.dart          buildAndroidWithGradle
package:flutter_tools/src/commands/build_apk.dart 620      buildApk.<async>
===== asynchronous gap ===========================
dart:async                                                 Future.Future.microtask
package:flutter_tools/src/commands/build_apk.dart          buildApk
package:flutter_tools/src/android/android_device.dart 287  AndroidDevice.startApp.<async>
dart:async                                                 _SyncCompleter.complete
package:flutter_tools/src/base/process.dart 113            runCommandAndStreamOutput.<async>
===== asynchronous gap ===========================
dart:async                                                 _asyncThenWrapperHelper
package:flutter_tools/src/android/android_device.dart      AndroidDevice.startApp
package:flutter_tools/src/run_hot.dart 160                 HotRunner._run.<async>
dart:async                                                 _SyncCompleter.complete
package:flutter_tools/src/resident_runner.dart 184         ResidentRunner.startEchoingDeviceLog.<async>
===== asynchronous gap ===========================
dart:async                                                 _asyncThenWrapperHelper
package:flutter_tools/src/run_hot.dart                     HotRunner._run
package:flutter_tools/src/run_hot.dart 77                  HotRunner.run.<fn>
package:stack_trace                                        Chain.capture
package:flutter_tools/src/run_hot.dart 76                  HotRunner.run
package:flutter_tools/src/commands/run.dart 278            RunCommand.runCommand.<async>

^C

### LINKED ISSUE #8148: gradle wrapper error when doing 'flutter run' on a fresh flutter project
## Steps to Reproduce

`flutter create` and then `flutter run` in the created directory.

## Logs

```
youngseokyoon list_view$ flutter run
Launching lib/main.dart on Pixel C in debug mode...
Running 'gradle wrapper'...                               

FAILURE: Build failed with an exception.

* Where:
Build file '/usr/local/google/home/youngseokyoon/Programming/temp/list_view/android/app/build.gradle' line: 14

* What went wrong:
A problem occurred evaluating project ':app'.
> Failed to apply plugin [id 'com.android.application']
   > Minimum supported Gradle version is 2.14.1. Current version is 2.13. If using the gradle wrapper, try editing the distributionUrl in /usr/local/google/home/youngseokyoon/Programming/temp/list_view/android/gradle/wrapper/gradle-wrapper.properties to gradle-2.14.1-all.zip

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
Exception from flutter run: Exception: Exception: Gradle failed: 1
package:flutter_tools/src/base/common.dart 29              throwToolExit
package:flutter_tools/src/android/gradle.dart 163          ensureGradlew.<async>
dart:async                                                 _SyncCompleter.complete
package:flutter_tools/src/base/process.dart 113            runCommandAndStreamOutput.<async>
===== asynchronous gap ===========================
dart:async                                                 _Completer.completeError
package:flutter_tools/src/android/gradle.dart 174          ensureGradlew.<async>
dart:async                                                 _SyncCompleter.complete
package:flutter_tools/src/base/process.dart 113            runCommandAndStreamOutput.<async>
===== asynchronous gap ===========================
dart:async                                                 _asyncThenWrapperHelper
package:flutter_tools/src/android/gradle.dart              ensureGradlew
package:flutter_tools/src/android/gradle.dart 192          buildGradleProject.<async>
===== asynchronous gap ===========================
dart:async                                                 Future.Future.microtask
package:flutter_tools/src/android/gradle.dart              buildGradleProject
package:flutter_tools/src/commands/build_apk.dart 610      buildAndroidWithGradle.<async>
===== asynchronous gap ===========================
dart:async                                                 Future.Future.microtask
package:flutter_tools/src/commands/build_apk.dart          buildAndroidWithGradle
package:flutter_tools/src/commands/build_apk.dart 620      buildApk.<async>
===== asynchronous gap ===========================
dart:async                                                 Future.Future.microtask
package:flutter_tools/src/commands/build_apk.dart          buildApk
package:flutter_tools/src/android/android_device.dart 287  AndroidDevice.startApp.<async>
dart:async                                                 _SyncCompleter.complete
package:flutter_tools/src/base/process.dart 113            runCommandAndStreamOutput.<async>
===== asynchronous gap ===========================
dart:async                                                 _asyncThenWrapperHelper
package:flutter_tools/src/android/android_device.dart      AndroidDevice.startApp
package:flutter_tools/src/run_hot.dart 160                 HotRunner._run.<async>
dart:async                                                 _SyncCompleter.complete
package:flutter_tools/src/resident_runner.dart 184         ResidentRunner.startEchoingDeviceLog.<async>
===== asynchronous gap ===========================
dart:async                                                 _asyncThenWrapperHelper
package:flutter_tools/src/run_hot.dart                     HotRunner._run
package:flutter_tools/src/run_hot.dart 77                  HotRunner.run.<fn>
package:stack_trace                                        Chain.capture
package:flutter_tools/src/run_hot.dart 76                  HotRunner.run
package:flutter_tools/src/commands/run.dart 278            RunCommand.runCommand.<async>

```

## Flutter Doctor

```
youngseokyoon list_view$ flutter doctor
[✓] Flutter (on Linux, channel master)
    • Flutter at /usr/local/google/home/youngseokyoon/Programming/fuchsia/lib/flutter
    • Framework revision 36f6405179 (2 hours ago), 2017-02-14 10:58:33
    • Engine revision f306a6f701
    • Tools Dart version 1.22.0-dev.10.6

[✓] Android toolchain - develop for Android devices (Android SDK 23.0.2)
    • Android SDK at /usr/local/google/home/youngseokyoon/Android/Sdk
    • Platform android-24, build-tools 23.0.2
    • ANDROID_HOME = /usr/local/google/home/youngseokyoon/Android/Sdk
    • OpenJDK Runtime Environment (build 1.8.0_112-google-v7-146844476-143772575)

[-] IntelliJ IDEA Community Edition (version 2016.3)
    • Dart plugin not installed; this adds Dart specific functionality.
    • Flutter plugin not installed; this adds Flutter specific functionality.
    • For information about managing plugins, see
      https://www.jetbrains.com/help/idea/managing-plugins.html

[✓] Connected devices
    • Pixel C • 5C08000423 • android-arm • Android 7.0 (API 24)
```

### PR DESCRIPTION
**Pull request by jakobr-google**
The recommendation from Gradle is to check in the gradlew scripts and
contents of gradle/wrapper. This way we don't have to depend on the
system gradle from Android Studio to create the wrapper before we can
build.

Fixes #8148
Fixes #8131

### CODE REVIEW

File: packages/flutter_tools/templates/create/android.tmpl/gradlew [hunk 1]
```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env bash
```
Comment (abarth) [inline packages/flutter_tools/templates/create/android.tmpl/gradlew]: How is this file versioned?  Does gradle promise never to change this file?  Do we plan to never change the contents of this file in the template?

**Comment by abarth** [comment]
Where does the `.jar` file come from?  Does gradle promise to never change the contents of the jar?  Is the source of the jar available?

I'm very skeptical of this approach.  It seems like we're committing to too many implementation details in this template.

File: packages/flutter_tools/templates/create/android.tmpl/gradlew (hunk 1 above)
Comment (jakobr-google) [inline packages/flutter_tools/templates/create/android.tmpl/gradlew]: We only need to change this file if we need a new version of the gradle wrapper (not Gradle itself), which should be very rare. The only purpose of the wrapper is to download the right version of Gradle.

**Comment by jakobr-google** [comment]
According to https://docs.gradle.org/current/userguide/gradle_wrapper.html, adding these files is the right approach when using the wrapper. The only other alternative is to have `flutter doctor` make sure we have a new enough gradle installed, and then there's really no reason to use the wrapper also.

**Comment by abarth** [comment]
What is the problem with depending on the gradle from Android Studio?

**Comment by jakobr-google** [comment]
And the source is here: https://github.com/gradle/gradle/tree/master/subprojects/wrapper

**Comment by jakobr-google** [comment]
Only problem is that we then depend on having Android Studio installed. I'm fine with that approach, too, but then we shouldn't use the wrapper. And we'll need someone to add a check to `flutter doctor`.

**Comment by abarth** [comment]
We've already decided to take a dependency on having Android Studio installed.  For all practical purposes, having Android Studio installed is required for building real Android applications.

**Comment by abarth** [comment]
The first step in https://flutter.io/setup/#android-setup is to install Android Studio and has been for a long time.

**Comment by jakobr-google** [comment]
Alright. It just sounded like this was a big breaking issue yesterday. In that case, I'll abandon this PR. Who should I assign the flutter doctor issue to?

**Comment by mit-mit** [comment]
Actually, should we differentiate the requirements for what is needed to *build* the all versus what is needed to *develop* the app. It's totally reasonable to require Android Studio for the latter, but seems undesirable for the former. For example, I don't think a Travis job should require Android Studio!?

**Comment by mit-mit** [comment]
Resolved in chat: We will leave the default app template without the gradle wrapper. If a developer desires to use it (for example for building on a bot), they can simply add it themselves (which also gives them full control over the version, etc.)

**Comment by abarth** [comment]
> Who should I assign the flutter doctor issue to?

flutter doctor should be pretty hackable.  You might want to try adding the check yourself.  Otherwise, we can get someone in MTV to take a look at it.

**Comment by jakobr-google** [comment]
I'll take a look.

**Pull request closed without merging**